test-apps/display-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/MainActivity.kt

```diff
@@ -115,20 +115,42 @@ class AssetsPathHandler(context: Context) : WebViewAssetLoader.PathHandler {
     }
 }
 
+@Suppress("SpellCheckingInspection")
 class MainActivity : AppCompatActivity() {
     private lateinit var host: IModelJsHost
     private var promiseName: String = ""
+    private lateinit var env: JSONObject
+
+    companion object {
+        const val BIM_CACHE_DIR = "bim_cache"
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         WebView.setWebContentsDebuggingEnabled(true)
         val alwaysExtractAssets = true // for debugging, otherwise the host will only extract when app version changes
         host = IModelJsHost(this, alwaysExtractAssets, true)
         host.startup()
-        val frontend = object : MobileFrontend(host, "&standalone=true") {
+
+        var args = "&standalone=true"
+        loadEnvJson()
+        env.optString("IMJS_STANDALONE_FILENAME", "").takeIf { it.isNotEmpty() }?.let { fileName ->
+            // ensure fileName already exists in the external files
+            getExternalFilesDir(BIM_CACHE_DIR)?.let { filesDir ->
+                File(filesDir, fileName).takeIf { it.exists() }?.let {
+                    args += "&iModelName=${Uri.encode(it.toString())}"
+                }
+            }
+        }
+
+        if (env.has("IMJS_IGNORE_CACHE"))
+            args += "&ignoreCache=true"
+
+        val frontend = object : MobileFrontend(host, args) {
             override fun supplyEntryPoint(): String {
-                // If you want to connect to a local dev server instead of the built-in frontend, return something like: "192.168.86.20:3000"
-                return "https://${WebViewAssetLoader.DEFAULT_DOMAIN}/assets/frontend/index.html"
+                // Connect to a local dev server if specified in the env JSON, otherwise use the embedded frontend
+                return env.optString("IMJS_DEBUG_URL", "").takeIf { it.isNotEmpty() } ?:
```

Comment:
I wonder if optNonEmptyString should be added as an extension.

Reply:
Done